src/Silicon.cpp

```diff
@@ -874,124 +1135,493 @@ namespace galsim {
+
+        // if no wavelengths allocated, photonsWavelength will be null, but some
+        // compilers don't like mapping a null pointer to the GPU, so assign it to
+        // something safe instead. (It will never be accessed in this case so the
+        // content of what it points to doesn't matter).
+        if (!photonsHasAllocatedWavelengths) {
+            photonsWavelength = photonsX;
+        }
+        // same with the angles
+        if (!photonsHasAllocatedAngles) {
+            photonsDXDZ = photonsX;
+            photonsDYDZ = photonsY;
+        }
+        
+	// random arrays
+	double* diffStepRandomArray = diffStepRandom.data();
+	double* pixelNotFoundRandomArray = pixelNotFoundRandom.data();
+	double* conversionDepthRandomArray = conversionDepthRandom.data();
+	
+	// delta image
+	int deltaXMin = _delta.getXMin();
+	int deltaYMin = _delta.getYMin();
+        int deltaXMax = _delta.getXMax();
+        int deltaYMax = _delta.getYMax();
+	int deltaStep = _delta.getStep();
+	int deltaStride = _delta.getStride();
+
+	int emptypolySize = _emptypoly.size();
+
+        double* deltaData = _delta.getData();
+        Bounds<double>* pixelInnerBoundsData = _pixelInnerBounds.data();
+        Bounds<double>* pixelOuterBoundsData = _pixelOuterBounds.data();
+        Position<float>* horizontalBoundaryPointsData = _horizontalBoundaryPoints.data();
+        Position<float>* verticalBoundaryPointsData = _verticalBoundaryPoints.data();
+
+        double* abs_length_table_data = _abs_length_table_GPU.data();
+        
+        Position<double>* emptypolyData = _emptypolyGPU.data();
+
 #ifdef _OPENMP
-#pragma omp parallel reduction (+:addedFlux)
-        {
-#pragma omp for
+#ifndef GALSIM_USE_GPU
+#pragma omp parallel for reduction(+:addedFlux)
+#else
+#pragma omp target teams distribute parallel for map(to: photonsX[i1:i2-i1], photonsY[i1:i2-i1], photonsDXDZ[i1:i2-i1], photonsDYDZ[i1:i2-i1], photonsFlux[i1:i2-i1], photonsWavelength[i1:i2-i1], diffStepRandomArray[0:(i2-i1)*2], conversionDepthRandomArray[0:i2-i1], pixelNotFoundRandomArray[0:i2-i1]) reduction(+:addedFlux)
 #endif
-        for (int i=i1; i<i2; i++) {
-            // Get the location where the photon strikes the silicon:
-            double x0 = photons.getX(i); // in pixels
-            double y0 = photons.getY(i); // in pixels
-            xdbg<<"x0,y0 = "<<x0<<','<<y0;
-
-            double dz = calculateConversionDepth(photons, i, conversionDepthRandom[i-i1]);
-            if (photons.hasAllocatedAngles()) {
-                double dxdz = photons.getDXDZ(i);
-                double dydz = photons.getDYDZ(i);
-                double dz_pixel = dz * invPixelSize;
-                x0 += dxdz * dz_pixel; // dx in pixels
-                y0 += dydz * dz_pixel; // dy in pixels
-            }
-            xdbg<<" => "<<x0<<','<<y0;
-            // This is the reverse of depth. zconv is how far above the substrate the e- converts.
```

Comment:
You got rid of most of these inline comments in your refactoring.  Please add them back.